cluster-autoscaler/core/loop_state.go

```diff
@@ -0,0 +1,448 @@
+	if err != nil {
+		glog.Errorf("Failed to list all nodes: %v", err)
+		return ls.onError(errors.ApiCallError, err)
+	}
+	if len(ls.allNodes) == 0 {
+		return ls.onEmptyCluster("nodes", true)
+	}
+
+	ls.readyNodes, err = ls.autoscaler.ReadyNodeLister().List()
+	if err != nil {
+		glog.Errorf("Failed to list ready nodes: %v", err)
+		return ls.onError(errors.ApiCallError, err)
+	}
+
+	// Handle GPU case - allocatable GPU may be equal to 0 up to 15 minutes after
+	// node registers as ready. See https://github.com/kubernetes/kubernetes/issues/54959
+	// Treat those nodes as unready until GPU actually becomes available and let
+	// our normal handling for booting up nodes deal with this.
+	// TODO: Remove this call when we handle dynamically provisioned resources.
+	ls.allNodes, ls.readyNodes = gpu.FilterOutNodesWithUnreadyGpus(ls.allNodes, ls.readyNodes)
+	if len(ls.readyNodes) == 0 {
+		// Cluster Autoscaler may start running before nodes are ready.
+		// Timeout ensures no ClusterUnhealthy events are published immediately in this case.
+		emit := ls.currentTime.After(ls.autoscaler.startTime.Add(nodesNotReadyAfterStartTimeout))
+		return ls.onEmptyCluster("ready nodes", emit)
+	}
+	return ls.onSuccess()
+}
+
+func (ls *loopState) updateClusterStateRegistry() (ExecutionResult, errors.AutoscalerError) {
+	err := ls.autoscaler.ClusterStateRegistry.UpdateNodes(ls.allNodes, ls.currentTime)
+	if err != nil {
+		glog.Errorf("Failed to update node registry: %v", err)
+		ls.autoscaler.scaleDown.CleanUpUnneededNodes()
+		return ls.onError(errors.CloudProviderError, err)
+	}
+	UpdateClusterStateMetrics(ls.autoscaler.ClusterStateRegistry)
+
+	metrics.UpdateDurationFromStart(metrics.UpdateState, ls.startTime)
+	metrics.UpdateLastTime(metrics.Autoscaling, time.Now())
+
+	return ls.onSuccess()
+}
+
+func (ls *loopState) cleanUnregisteredNodes() (ExecutionResult, errors.AutoscalerError) {
+	// Check if there are any nodes that failed to register in Kubernetes master.
+	unregisteredNodes := ls.autoscaler.ClusterStateRegistry.GetUnregisteredNodes()
+	if len(unregisteredNodes) > 0 {
+		glog.V(1).Infof("%d unregistered nodes present", len(unregisteredNodes))
+		removedAny, err := removeOldUnregisteredNodes(unregisteredNodes, ls.context, ls.currentTime, ls.context.LogRecorder)
+		// There was a problem with removing unregistered nodes. Retry in the next loop.
+		if err != nil {
+			if removedAny {
+				glog.Warningf("Some unregistered nodes were removed, but got error: %v", err)
+			} else {
+				glog.Errorf("Failed to remove unregistered nodes: %v", err)
+			}
+			return ls.onError(errors.CloudProviderError, err)
+		}
+		// Some nodes were removed. Let's skip this iteration, the next one should be better.
+		if removedAny {
+			glog.V(0).Infof("Some unregistered nodes were removed, skipping iteration")
+			return ls.onEarlyExit()
+		}
+	}
+	return ls.onSuccess()
+}
+
+func (ls *loopState) checkClusterHealth() (ExecutionResult, errors.AutoscalerError) {
```

Comment:
It doesn't just "check" cluster health, but also act on this (cleaning up internal state and exiting early from the main loop)

Reply:
Yes, it's one of the many places where we have side-effects in innocent-looking places.
I don't see (yet) a good way to fix this particular issue.